javalib/src/main/scala/java/util/Random.scala

```diff
@@ -17,17 +18,60 @@ class Random(seed_in: Long) extends AnyRef with java.io.Serializable {
   def this() = this(Random.randomSeed())
 
   def setSeed(seed_in: Long): Unit = {
-    seed = (seed_in ^ 0x5DEECE66DL) & ((1L << 48) - 1)
+    val seed = ((seed_in ^ 0x5DEECE66DL) & ((1L << 48) - 1)) // as documented
+    seedHi = (seed >>> 24).toInt
+    seedLo = seed.toInt & ((1 << 24) - 1)
     haveNextNextGaussian = false
   }
 
   protected def next(bits: Int): Int = {
-    seed = (seed * 0x5DEECE66DL + 0xBL) & ((1L << 48) - 1)
-    (seed >>> (48 - bits)).toInt
+    /* This method is originally supposed to work with a Long seed from which
+     * 48 bits are used.
+     * Since Longs are too slow, we manually decompose the 48-bit seed in two
+     * parts of 24 bits each.
+     * The computation below is the translation in 24-by-24 bits of the
+     * specified computation, taking care never to produce intermediate values
+     * requiring more than 52 bits of precision.
+     */
+
+    @inline
+    def _24msbOf(x: Double): Int = (x / (1 << 24).toDouble).toInt
```

Comment:
If the difference of the exponent parts is in the -1023 to 1023 range, yes. In general, I'd there *exists* an exact solution, it is guaranteed that it will be found.